vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBoxDataController.java

```diff
@@ -0,0 +1,604 @@
+                .addPropertyChangeListener("opened", this::clearFilterOnClose);
+    }
+
+    /**
+     * Called to notify this controller that the component has been detached
+     */
+    void onDetach() {
+        if (dataProviderListener != null) {
+            dataProviderListener.remove();
+            dataProviderListener = null;
+        }
+
+        if (clearFilterOnCloseRegistration != null) {
+            clearFilterOnCloseRegistration.remove();
+            clearFilterOnCloseRegistration = null;
+        }
+    }
+
+    /**
+     * Refresh item data of the web component when data has been updated, or
+     * after changes that might affect the presentation / rendering of items
+     */
+    void reset() {
+        lastFilter = null;
+        if (dataCommunicator != null) {
+            dataCommunicator.setRequestedRange(0, 0);
+            dataCommunicator.reset();
+        }
+        comboBox.runBeforeClientResponse(ui -> ui.getPage().executeJs(
+                // If-statement is needed because on the first attach this
+                // JavaScript is called before initializing the connector.
+                "if($0.$connector) $0.$connector.reset();",
+                comboBox.getElement()));
+    }
+
+    /**
+     * Called to confirm that an update has been processed by the client-side
+     * connector
+     */
+    void confirmUpdate(int id) {
+        dataCommunicator.confirmUpdate(id);
+    }
+
+    /**
+     * Called when the client-side connector requests data
+     */
+    void setRequestedRange(int start, int length, String filter) {
+        dataCommunicator.setRequestedRange(start, length);
+        filterSlot.accept(filter);
+    }
+
+    /**
+     * Called by the client-side connector to reset the data communicator
+     */
+    void resetDataCommunicator() {
+        /*
+         * The client filter from combo box will be used in the data
+         * communicator only within 'setRequestedRange' calls to data provider,
+         * and then will be erased to not affect the data view item count
+         * handling methods. Thus, if the current client filter is not empty,
+         * then we need to re-set it in the data communicator.
+         */
+        if (lastFilter == null || lastFilter.isEmpty()) {
+            dataCommunicator.reset();
+        } else {
+            String filter = lastFilter;
+            lastFilter = null;
+            /*
+             * This filter slot will eventually call the filter consumer in data
+             * communicator and 'DataCommunicator::reset' is done inside this
+             * consumer, so we don't need to explicitly call it.
+             */
+            filterSlot.accept(filter);
+        }
+    }
+
+    // ****************************************************
+    // List data view implementation
+    // ****************************************************
+
+    @Override
+    public ComboBoxListDataView<TItem> getListDataView() {
```

Comment:
Decided to implement the `HasDataView`, `HasListDataView`, `HasLazyDataView` interfaces here as well. Some of the `setItems` methods call some of the default implementations of the interfaces, and it require more changes to sort this out. This also leads to these methods requiring public visibility. Might want to take a second look at this after merging.